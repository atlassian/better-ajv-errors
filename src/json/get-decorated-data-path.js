export default function getDecoratedDataPath(jsonAst, dataPath) {
  // TODO: Handle json pointer escape notation and better error handling
  const pointers = dataPath.split('/').slice(1);
  let decoratedPath = '';
  pointers.reduce((obj, pointer) => {
    switch (obj.type) {
      case 'Object': {
        decoratedPath += `/${pointer}`;
        const filtered = obj.children.filter(
          child => child.key.value === pointer
        );
        if (filtered.length !== 1) {
          throw new Error(`Couldn't find property ${pointer} of ${dataPath}`);
        }
        return filtered[0].value;
      }
      case 'Array':
        decoratedPath += `/${pointer}${getTypeName(obj.children[pointer])}`;
        return obj.children[pointer];
      default:
        // eslint-disable-next-line no-console
        console.log(obj);
    }
  }, jsonAst);
  return decoratedPath;
}

function getTypeName(obj) {
  if (!obj || !obj.children) {
    return '';
  }
  const type = obj.children.filter(child => child.key.value === 'type');

  if (!type.length) {
    return '';
  }

  return (type[0].value && `:${type[0].value.value}`) || '';
}
