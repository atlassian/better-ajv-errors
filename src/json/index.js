export { default as getMetaFromPath } from './get-meta-from-path';
export { default as getDecoratedDataPath } from './get-decorated-data-path';
