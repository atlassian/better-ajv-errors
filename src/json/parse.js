// Custom JSON parser based on https://github.com/douglascrockford/JSON-js/blob/master/json_parse.js
let line;
let column;
let at; // The index of the current character
let ch; // The current character

const META = Symbol('meta');
const VALUE = Symbol('value');

export const getValue = (obj, key) => obj[VALUE][key] || obj[VALUE][VALUE][key];
export const getMeta = obj => obj[META];

const escapee = {
  '"': '"',
  '\\': '\\',
  '/': '/',
  b: '\b',
  f: '\f',
  n: '\n',
  r: '\r',
  t: '\t',
};
let text;

const error = function(m) {
  // Call error when something is wrong.
  throw {
    name: 'SyntaxError',
    message: m,
    at: at,
    text: text,
  };
};

const next = function(c) {
  // If a c parameter is provided, verify that it matches the current character.

  if (c && c !== ch) {
    error("Expected '" + c + "' instead of '" + ch + "'");
  }

  // Get the next character. When there are no more characters,
  // return the empty string.

  ch = text.charAt(at);
  at += 1;
  if (ch === '\n') {
    line += 1;
    column = 0;
  } else {
    column += 1;
  }
  return ch;
};

const number = function() {
  // Parse a number value.

  let value;
  let string = '';

  if (ch === '-') {
    string = '-';
    next('-');
  }
  while (ch >= '0' && ch <= '9') {
    string += ch;
    next();
  }
  if (ch === '.') {
    string += '.';
    while (next() && ch >= '0' && ch <= '9') {
      string += ch;
    }
  }
  if (ch === 'e' || ch === 'E') {
    string += ch;
    next();
    if (ch === '-' || ch === '+') {
      string += ch;
      next();
    }
    while (ch >= '0' && ch <= '9') {
      string += ch;
      next();
    }
  }
  value = +string;
  if (!isFinite(value)) {
    error('Bad number');
  } else {
    return value;
  }
};

const string = function() {
  // Parse a string value.

  let hex;
  let i;
  let value = '';
  let uffff;

  // When parsing for string values, we must look for " and \ characters.

  if (ch === '"') {
    while (next()) {
      if (ch === '"') {
        next();
        return value;
      }
      if (ch === '\\') {
        next();
        if (ch === 'u') {
          uffff = 0;
          for (i = 0; i < 4; i += 1) {
            hex = parseInt(next(), 16);
            if (!isFinite(hex)) {
              break;
            }
            uffff = uffff * 16 + hex;
          }
          value += String.fromCharCode(uffff);
        } else if (typeof escapee[ch] === 'string') {
          value += escapee[ch];
        } else {
          break;
        }
      } else {
        value += ch;
      }
    }
  }
  error('Bad string');
};

const white = function() {
  // Skip whitespace.

  while (ch && ch <= ' ') {
    next();
  }
};

const word = function() {
  // true, false, or null.

  switch (ch) {
    case 't':
      next('t');
      next('r');
      next('u');
      next('e');
      return true;
    case 'f':
      next('f');
      next('a');
      next('l');
      next('s');
      next('e');
      return false;
    case 'n':
      next('n');
      next('u');
      next('l');
      next('l');
      return null;
  }
  error("Unexpected '" + ch + "'");
};

let value; // Place holder for the value function.

const array = function() {
  // Parse an array value.

  const arr = [];
  if (ch === '[') {
    next('[');
    white();
    if (ch === ']') {
      next(']');
      return arr; // empty array
    }
    while (ch) {
      arr.push(value());
      white();
      if (ch === ']') {
        next(']');
        return arr;
      }
      next(',');
      white();
    }
  }
  error('Bad array');
};

const object = function() {
  // Parse an object value.

  let key;
  const obj = {};
  obj[META] = { line, column };
  obj[VALUE] = {};

  if (ch === '{') {
    next('{');
    white();
    if (ch === '}') {
      next('}');
      return obj; // empty object
    }
    while (ch) {
      const c = column;
      key = string();
      white();
      next(':');
      if (Object.hasOwnProperty.call(obj, key)) {
        error("Duplicate key '" + key + "'");
      }
      obj[VALUE][key] = {};
      obj[VALUE][key][META] = { line, column: c };
      obj[VALUE][key][VALUE] = value();
      white();
      if (ch === '}') {
        next('}');
        return obj;
      }
      next(',');
      white();
    }
  }
  error('Bad object');
};

value = function() {
  // Parse a JSON value. It could be an object, an array, a string, a number,
  // or a word.

  white();
  switch (ch) {
    case '{':
      return object();
    case '[':
      return array();
    case '"':
      return string();
    case '-':
      return number();
    default:
      return ch >= '0' && ch <= '9' ? number() : word();
  }
};

// Return the json_parse function. It will have access to all of the above
// functions and variables.
export default source => {
  let result;

  text = source;
  at = 0;
  ch = ' ';
  line = 0;
  column = 0;
  result = value();
  white();
  if (ch) {
    error('Syntax error');
  }
  return result;
};
