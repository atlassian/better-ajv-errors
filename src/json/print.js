import chalk from 'chalk';
import { getMetaFromPath } from './';

export default (obj, dataPath, options = {}) => annotate => {
  const {
    delta = 5,
    defaultColor = 'gray',
    errorColor = 'red',
    annotationColor = 'cyan',
    gutterChar = 'â”‚',
    annotationGutterChar = 'ã€‰',
    indent = 2,
  } = options;

  try {
    const jsonString = JSON.stringify(obj, null, indent);
    const { line, column } = getMetaFromPath(jsonString, dataPath);

    const lines = jsonString.split('\n');
    const gutterWidth = lines.length.toString().length;
    const annotationGutter = `${''.padStart(
      gutterWidth
    )}${annotationGutterChar}`;

    const min = Math.max(line - delta, 0);
    const max = Math.min(line + delta, lines.length - 1);
    let output = [];

    for (let i = min; i <= max; ++i) {
      const color = i === line ? errorColor : defaultColor;
      output.push(
        chalk`{${color} ${String(i + 1).padStart(
          gutterWidth
        )}${gutterChar} ${lines[i]}}`
      );
      if (i === line) {
        output = output.concat([
          // New line
          chalk`{${annotationColor} ${annotationGutter}}`,

          chalk`{${annotationColor} ${annotationGutter}${''.padStart(
            column
          )}${annotate()}}`,

          // New line
          chalk`{${annotationColor} ${annotationGutter}}`,
        ]);
      }
    }
    return output;
  } catch (e) {
    return [chalk`{red ðŸ˜¢  ${e.message || 'Error'}}`];
  }
};
