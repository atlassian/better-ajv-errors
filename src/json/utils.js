// TODO: Better error handling
export const getPointers = dataPath => {
  return dataPath
    .split('/')
    .slice(1)
    .map(pointer => pointer.split('~1').join('/').split('~0').join('~'));
};
