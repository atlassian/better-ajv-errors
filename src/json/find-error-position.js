import parse, { META_SYMBOL, VALUE_SYMBOL } from './parse';

export default function findErrorPosition(jsonString, dataPath) {
  // console.log(JSON.stringify(parse(jsonString), null, 2));
  // TODO: Handle json pointer escape notation and better error handling
  const { [META_SYMBOL]: { line, column } } = dataPath
    .split('/')
    .slice(1)
    .reduce(
      (obj, pointer) => obj[VALUE_SYMBOL][pointer] || obj[VALUE_SYMBOL][VALUE_SYMBOL][pointer],
      parse(jsonString)
    );

  return { line, column };
}
