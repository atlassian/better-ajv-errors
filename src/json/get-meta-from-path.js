import { getPointers } from './utils';

export default function getMetaFromPath(
  jsonAst,
  dataPath,
  isIdentifierLocation
) {
  const pointers = getPointers(dataPath);
  const lastPointerIndex = pointers.length - 1;
  return pointers.reduce((obj, pointer, idx) => {
    switch (obj.type) {
      case 'Object': {
        const filtered = obj.children.filter(
          child => child.key.value === pointer
        );
        if (filtered.length !== 1) {
          throw new Error(`Couldn't find property ${pointer} of ${dataPath}`);
        }
        const { key, value } = filtered[0];
        return isIdentifierLocation && idx === lastPointerIndex ? key : value;
      }
      case 'Array':
        return obj.children[pointer];
      default:
        // eslint-disable-next-line no-console
        console.log(obj);
    }
  }, jsonAst);
}
