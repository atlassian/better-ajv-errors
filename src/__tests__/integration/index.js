import * as adfSchema from './adfSchema.json';
import Ajv from 'ajv';
import betterAjvErrors from '../..';
import fs from 'fs';

const ajv = new Ajv({ jsonPointers: true, schemaId: 'auto' });
ajv.addMetaSchema(require('ajv/lib/refs/json-schema-draft-04.json'));
const validate = ajv.compile(adfSchema);

const readFilesSync = path => {
  return fs.readdirSync(path).reduce((acc, name) => {
    if (name.match(/\.json$/)) {
      acc.push({
        name,
        data: JSON.parse(fs.readFileSync(`${path}/${name}`, 'utf-8')),
      });
    }

    return acc;
  }, []);
};

describe('Integration Test - ', () => {
  const invalid = readFilesSync(`${__dirname}/invalid`);
  invalid.forEach(file => {
    it(`shows better error for '${file.name}`, () => {
      const valid = validate(file.data);
      expect(valid).toBeFalsy();
      const output = betterAjvErrors(adfSchema, file.data, validate.errors, {
        format: 'js',
      });
      expect(output).toMatchSnapshot();
    });
  });
});
