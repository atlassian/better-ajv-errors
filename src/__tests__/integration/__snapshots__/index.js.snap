// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration Test -  shows better error for 'link-without-attrs.json 1`] = `
Array [
  Object {
    "error": "/content/0:paragraph/content/0:text/marks/0:link should have required property 'attrs'",
    "path": "/content/0/content/0/marks/0",
    "start": Object {
      "column": 106,
      "line": 1,
      "offset": 105,
    },
  },
]
`;
